### ISSUE TITLE
Define `pelkmanslab`/`fmi` parsl config

### ISSUE THREAD

**Issue by tcompa** [fractal-analytics-platform/fractal-server#34]
After https://github.com/fractal-analytics-platform/fractal-server/pull/24 and https://github.com/fractal-analytics-platform/fractal-server/issues/4, the current `pelkmanslab` config only has three executors:
* `cpu`
* `cpu-2` (only useful for testing)
* `gpu`

We should implement the list we discussed about (low/high mem/cpu). And then the same should be done for some FMI machine.

**Comment by tcompa** [comment]
This also includes:
>  What is a typical SLURM partition name at the FMI machines?

**Comment by jluethi** [comment]
Regarding relevant executors, we should have a look at memory & CPU profiling from some example jobs. I imagine once things run somewhat memory efficient, that depends on ROI sizes, but we could at least get a general understanding of how different tasks have fairly different requirements.
I'd expect the parsing tasks (yokogawa_to_zarr) to be very memory efficient if we ensure every image is parsed & saved to disk (assuming they are all in different chunks, which they are for us), maybe something in the environment of 2GB of memory. Also, probably doesn't parallelize much, so 1-2 CPUs.

Segmentation is a very different story and may often require much more memory (and maybe also parallel computation)?

Given that everything scales with ROI sizes, we may offer a selection of typical flavors (to be tested & checked with monitoring):
- cpu-low: 1 CPU, 2 GB of memory
- cpu-intermediate: 4 CPUs, 16 GB of memory
- cpu-high: 16 CPUs, 64 GB of memory
- cpu-very-high: 32 CPUs, 128 GB of memory (or some other very large case on the FMI side, this is inspired by UZH node sizes)

On the GPU side, if UZH is anyway exclusive on node usage, it can just be the full node for the moment. Not sure yet what will make sense on the FMI side.

Potential matching:
Parsing: CPU-low
Illum-corr: CPU-low or CPU-intermediate
Segmentation: CPU-high or GPU
Measurements: CPU-intermediate or even CPU-high in some cases

cc @gusqgm for "What is a typical SLURM partition name at the FMI machines?" => what was your setup you had with the legacy Fractal version?

**Comment by jluethi** [comment]
@tcompa for the FMI side:
The SLURM partition on the test cluster is `cpu_long`. It's a single node with 56 cores and 184 GB of RAM.

We could go with a similar setup for CPU executors here:
cpu-low: 1 CPU, 4 GB of memory
cpu-intermediate: 4 CPUs, 16 GB of memory
cpu-high: 16 CPUs, 64 GB of memory
cpu-very-high: 50 CPUs, 180 GB of memory (or some other very large case on the FMI side, this is inspired by UZH node sizes)

Regarding GPU executors: Those are not available on the test cluster. On the main cluster, they will be on different slurm partitions (the main cluster has ~6 slurm partitions or so, different GPU & CPU ones (some with longer and some with shorter runtimes).

**Comment by tcompa** [comment]
Thanks for all the details above.

What should be the maximum number of active jobs? (on the different partitions)
This is the number `N` such that if `N>num_wells`, then all wells will be processed in parallel.

**Comment by jluethi** [comment]
@tcompa I wouldn't limit this much. Let's see performance a bit, we may hit the point where things are mostly IO limited. But otherwise, there is nothing that would say "don't parse 23 well at once", or even "don't parse 100 wells at once".
If we need to set an N, let's go with 100 for the moment? And then slurm decides when the jobs get priority.

On the GPU side, this limit would be useful. i.e. don't run more than 4 GPU jobs for a workflow for the time being, until we figured out either how to run multiple jobs on the same GPU node or such a thing.

**Comment by tcompa** [comment]
Self reminder:
- [ ] Do we want to have one slurm job per fractal task (as it was in legacy)? To be assessed.

Ref: https://github.com/fractal-analytics-platform/fractal-tasks-core/issues/30

**Comment by jluethi** [comment]
>  Do we want to have one slurm job per fractal task (as it was in legacy)? To be assessed.

This discussion clarified a few things for me. I was so far thinking about slurm jobs & fractal tasks cpu & memory specification as the same thing. And an easy way to have direct control may be to move back to that model.

But given this logic, could we also move to a system where we do more the following:
A user sets some parameters for their tasks regarding CPU, GPU & memory. Fractal collects all these information and submits slurm jobs that can have 1-N tasks, choosing a fitting executor.

e.g. if the user has 20 wells, each needs 1 CPU & 3.5 GB of GPU, we could submit them in 2 executors with 16 CPUs & 60 GB of RAM each. If the user submits 24 wells, that still fits in that envelope. If the submit 36 wells, maybe we use 2x 16 core executor and 1 4 core? Or just an additional executor.

In that way, we could also be very flexible in what users can specify as memory need and through parsl, we just bundle them into packages that together don't exceed one of our standard executors. In that way though, the user would never specify the executor, just the parameters of their job

**Comment by tcompa** [comment]
> > Do we want to have one slurm job per fractal task (as it was in legacy)? To be assessed.
> 
> This discussion clarified a few things for me. I was so far thinking about slurm jobs & fractal tasks cpu & memory specification as the same thing. And an easy way to have direct control may be to move back to that model.

I think this was the behavior in legacy. I'll see if we can switch back to that behavior, for clarity and for the purpose of running tests for the demo.
In principle this may also allow your other suggestion of running several low-resources tasks on the same node (it just requires `exclusive=False`), but let's think about it a bit later once we are happy with the first version.

**Comment by tcompa** [comment]
Again, for the record:

* I started an example for parsing images from a single 9x8 well (19 Z planes and 3 channels).
* The 0-th level (i.e. a zarr folder of 22G) is computed and written to disk in ~35 minutes.
* The complete task (including writing the 5-levels pyramid) takes ~60 minutes and writes ~30 G to disk.

[this is all with minimal requirements (1 CPU and 7G - although actual memory usage is much smaller)]

**Comment by jluethi** [comment]
That's good to hear! Once we have the large dataset running, let's see how it scales when multiple wells are run in parallel. If it's somewhere in the same order of magnitude (i.e. ~2-3h for the whole thing), we're similar to the fractal monolith version :) Seems reasonable in general what we're getting here.

**Comment by tcompa** [comment]
Caveat for the future (especially for the `fmi` install): see https://github.com/fractal-analytics-platform/fractal-server/issues/76#issuecomment-1257825858 for an issue about the total memory that can be requested in the `gpu` partition.

TL;DR
SLURM has intrinsic memory limits which are not completely transparent (as they come from an internal evaluation based on both memory and other job requirements). Sometimes just reducing the request by 1G solves a problem.

**Comment by jluethi** [comment]
> TL;DR
SLURM has intrinsic memory limits which are not completely transparent (as they come from an internal evaluation based on both memory and other job requirements). Sometimes just reducing the request by 1G solves a problem.

@tcompa Yes, just ran into this issue again :) While 63 GB can run on some CPU nodes on the pelkmans lab front, it's only the 128 GB nodes. Thus, the cpu-high executor should be adjusted on the pelkmans lab infrastructure down to ~60 GB? (maybe 62 would still run?). While those nodes should have 64 GB, requesting 64 or 63 appears to be too much, we need to go slightly lower to actually get those nodes.

With the current cpu-high, we can only run max. 6 jobs of those in parallel (currently running into this with the 23 well case :)). Plus, we can only run 1 cpu-high job on the huge nodes, even though they should fit 2 jobs with 60 GB memory
<img width="1032" alt="Bildschirmfoto 2022-09-28 um 08 29 53" src="https://user-images.githubusercontent.com/18033446/192744719-55490ced-42cc-4469-8ea6-c2f4c45aec99.png">

**Comment by tcompa** [comment]
Ref https://github.com/fractal-analytics-platform/fractal-server/pull/85
